### PR TITLE
switch from states mapping to two state variables for the genesis and…

### LINKED ISSUE #1938: simplify light client contract: create genesis state struct and remove states mapping
subtask of https://github.com/EspressoSystems/espresso-sequencer/issues/1912
1. remove unnecessary variables from the `LightClientState`
```
struct LightClientState {
        uint64 viewNum;
        uint64 blockHeight;
        BN254.ScalarField blockCommRoot;
}
```
2. introduce a new struct for genesis state
```
struct GenesisState {
        uint256 threshold;
        uint256 stakeTableBlsKeyComm;
        uint256 stakeTableSchnorrKeyComm;
        uint256 stakeTableAmountComm;
        uint256 blockCommRoot;
 }
```
To ensure simplicity for the `updateStateHistory(uint64 blockNumber, uint64 blockTimestamp, LightClientState memory state)` function, there may be need for the genesisState to be separated into two separate variables e.g.
```
struct LightClientState {
        uint64 viewNum;
        uint64 blockHeight;
        BN254.ScalarField blockCommRoot;
}

struct StakeState {
        uint256 threshold;
        BN254.ScalarField stakeTableBlsKeyComm;
        BN254.ScalarField stakeTableSchnorrKeyComm;
        BN254.ScalarField stakeTableAmountComm;
}
   
StakeStake genesisStake;
LightClientState genesisState;
LightClientState lightClientState;
```
3. remove the states mapping and  add a `genesisState` variable and add a `finalizedState` variable

### PR DESCRIPTION
**Pull request by alysiahuggins**
… finalized state

Closes #1938 

### This PR:
- removes states mapping
- added global variable for genesisState and finalizedState
- WIP

### CODE REVIEW

**Comment by philippecamacho** [comment]
Made some progress fixing the compilation errors but not there yet:
* Created a new [branch](https://github.com/EspressoSystems/HotShot/tree/lc-contract-updates) of HotShot (fork of tag = "0.5.72") and added the `StakeState` struct.
* Have the espresso-sequencer repository point to this branch
* Fix several compilation errors.

Right now, for some reason the espresso-types crates does not compile.

**Comment by alysiahuggins** [comment]
feature branch, `lc-contract-updates` have been added to the following repos so that we can support the new types introduced by this ticket:
- https://github.com/EspressoSystems/HotShot
- https://github.com/EspressoSystems/hotshot-builder-core
- https://github.com/EspressoSystems/marketplace-builder-core
- https://github.com/EspressoSystems/hotshot-events-service
- https://github.com/EspressoSystems/hotshot-query-service
https://github.com/EspressoSystems/espresso-sequencer/commit/d391f7e16953566b229245ec65713196616e6650

**Comment by alysiahuggins** [comment]
- removed fields from the `LightClientState` struct so it now looks like this 
```
struct LightClientState {
        uint64 viewNum;
        uint64 blockHeight;
        BN254.ScalarField blockCommRoot;
    }
```
- updated related rust code such as the lightclient `adapter` and `MockGenesis` action in the `diff-tests`, to continue, one would have to also update the `MockConsecutiveFinalizedStates`, `MockSkipBlocks`, `MockMissEndingBlock` etc


https://github.com/EspressoSystems/espresso-sequencer/pull/1939/commits/3a27dda9fce8ab076fbd629ea9b4e327e9f39c2c
this git commit ignored verification checks so that anyone can pick up from where i left off.

File: contracts/src/LightClient.sol [hunk 1]
```diff
@@ -91,20 +94,22 @@ contract LightClient is Initializable, OwnableUpgradeable, UUPSUpgradeable {
...
+    /// @notice The finalized HotShot Stake state (as the digest of the entire HotShot state)
+    /// @param threshold The (stake-weighted) quorum threshold for a QC to be considered as valid
+    /// @param stakeTableBlsKeyComm The commitment to the BlsVerKey column of the stake table
+    /// @param stakeTableSchnorrKeyComm The commitment to the SchnorrVerKey column of the table
+    /// @param stakeTableAmountComm The commitment to the stake amount column of the stake table
+    struct StakeState {
+        uint256 threshold;
         BN254.ScalarField stakeTableBlsKeyComm;
         BN254.ScalarField stakeTableSchnorrKeyComm;
         BN254.ScalarField stakeTableAmountComm;
```
Comment (alxiong) [inline contracts/src/LightClient.sol:112]: naming nit:
```suggestion
    struct StakeTableState {
        uint256 threshold;
        BN254.ScalarField blsKeyComm;
        BN254.ScalarField schnorrKeyComm;
        BN254.ScalarField amountComm;
```

reason: 
- `StakeTableState` is clearer imo
- once we have the enclosing struct name to have the "stake table" prefix, we don't need it in the names of its fields, because `stakeTableState.blsKeyComm` already provides a clear context.

File: contracts/src/LightClient.sol [hunk 2]
```diff
@@ -62,6 +56,15 @@ contract LightClient is Initializable, OwnableUpgradeable, UUPSUpgradeable {
     /// @notice mapping to store light client states in order to simplify upgrades
     mapping(uint32 index => LightClientState value) public states;
```
Comment (alxiong) [inline contracts/src/LightClient.sol:57]:
```suggestion
```

we should remove this by now?

File: contracts/src/LightClient.sol [hunk 3]
```diff
@@ -199,34 +203,38 @@ contract LightClient is Initializable, OwnableUpgradeable, UUPSUpgradeable {
...
         stateHistoryRetentionPeriod = _stateHistoryRetentionPeriod;
 
-        bytes32 initStakeTableComm = computeStakeTableComm(_genesis);
+        bytes32 initStakeTableComm = computeStakeTableComm(_genesisStakeState);
         votingStakeTableCommitment = initStakeTableComm;
-        votingThreshold = _genesis.threshold;
+        votingThreshold = _genesisStakeState.threshold;
         frozenStakeTableCommitment = initStakeTableComm;
-        frozenThreshold = _genesis.threshold;
+        frozenThreshold = _genesisStakeState.threshold;
```
Comment (alxiong) [inline contracts/src/LightClient.sol:237]: imo, we should remove all of these: 
```solidity
bytes32 public votingStakeTableCommitment;
uint256 public votingThreshold;
bytes32 public frozenStakeTableCommitment;
uint256 public frozenThreshold;
mapping(uint32 index => LightClientState value) public states;

function computeStakeTableComm(LightClientState memory state) returns (bytes32);
```

i roughly explain the "frozen/voting stake table" here: https://www.notion.so/espressosys/Light-Client-Contract-Details-6c396f7824e34356bdcda710015291f9?pvs=4#57affaec4dbf4508a27b169255fde936

TL;DR is `votingXX` is the stake table snapshot/version that we will be using when validating quorum signatures (or SNARK that proves a valid QC)

but since we are not changing them anyway, just remove them for now and use `genesisState` you introduced.

File: contracts/src/LightClient.sol [hunk 4]
```diff
@@ -282,12 +285,17 @@ contract LightClient is Initializable, OwnableUpgradeable, UUPSUpgradeable {
...
+    /// @dev Simple getter function for the genesis stake state
+    function getGenesisStakeState() public view virtual returns (StakeState memory) {
+        return genesisStakeState;
     }
 
     /// @dev Simple getter function for the finalized state
     function getFinalizedState() public view virtual returns (LightClientState memory) {
-        return states[finalizedState];
+        return finalizedState;
     }
```
Comment (alxiong) [inline contracts/src/LightClient.sol:299]: remove all of these helpers because these states are trivial to read (and come with default getter)

File: hotshot-state-prover/src/mock_ledger.rs [hunk 5]
```diff
@@ -347,21 +362,16 @@ impl MockLedger {
...
-            self.state.stake_table_comm.1,
-            self.state.stake_table_comm.2,
         ];
         let pi: GenericPublicInput<F> = pi.into();
         pi.into()
     }
 
     /// Returns the (bytes32 votingStakeTableComm, bytes32 frozenStakeTableComm) used in contract
-    pub fn get_stake_table_comms(&self) -> (H256, H256) {
+    pub fn get_stake_table_comms(&self) -> (H256, H256, ParsedStakeState) {
```
Comment (alxiong) [inline hotshot-state-prover/src/mock_ledger.rs:374]: is there a reason why we still return the first two commitments?
maybe renamed to `get_stake_table_state(&self) -> ParsedStakeTableState`

File: hotshot-state-prover/src/mock_ledger.rs [hunk 6]
```diff
@@ -99,14 +102,14 @@ impl MockLedger {
...
+        // let fee_ledger_comm: ark_ff::Fp<ark_ff::MontBackend<ark_bn254::FrConfig, 4>, 4> = F::from(5678);
 
         let genesis = LightClientState {
             view_number: 0,
             block_height: 0,
             block_comm_root,
-            fee_ledger_comm,
-            stake_table_comm: st.commitment(SnapshotVersion::LastEpochStart).unwrap(),
+            // fee_ledger_comm,
+            // stake_table_comm: st.commitment(SnapshotVersion::LastEpochStart).unwrap(),
```
Comment (alxiong) [inline hotshot-state-prover/src/mock_ledger.rs:112]: just a reminder to remove all commented lines later. no need to address it now.

File: hotshot-state-prover/src/circuit.rs
Comment (alxiong) [inline hotshot-state-prover/src/circuit.rs:1]: needs to update the code comments (and docs if any) to reflect the changes.

File: hotshot-state-prover/src/service.rs [hunk 7]
```diff
@@ -200,6 +200,35 @@ pub fn light_client_genesis_from_stake_table(
...
+    let st = init_stake_table_from_sequencer(sequencer_url, stake_table_capacity)
+        .await
+        .with_context(|| "Failed to initialize stake table")?;
+    light_client_genesis_stake_from_stake_table(st)
+}
+
+#[inline]
+pub fn light_client_genesis_stake_from_stake_table(
+    st: StakeTable<BLSPubKey, StateVerKey, CircuitField>,
+) -> anyhow::Result<ParsedStakeState> {
```
Comment (alxiong) [inline hotshot-state-prover/src/service.rs:216]: shouldn't we use `StakeTable.into()` I thought `From<hotshot_stake_table::StakeTable> for ParsedStakeTableState` is implemented in `adapters/light_client.rs`?

if not, we should add it there; and we can remove this function.

**Review by alxiong — COMMENTED** [review]
I've looked at `adapter/*`, `diff-test` , `hotshot-state-prover`
they all look very good to me. Alysia is doing great work. 👏

**Comment by alxiong** [comment]
@mrain can you also take a look at Alysia's changes to the `circuit.rs`. The changes look good to me at least. 

for reference, we are simplifying light client state as specified in https://github.com/EspressoSystems/espresso-sequencer/issues/1938

thanks!

**Comment by mrain** [comment]
> @mrain can you also take a look at Alysia's changes to the `circuit.rs`. The changes look good to me at least.
> 
> for reference, we are simplifying light client state as specified in #1938
> 
> thanks!

LGTM. Changes are simple enough.

File: contracts/rust/diff-test/src/main.rs [hunk 8]
```diff
@@ -503,9 +512,17 @@ fn main() {
...
                 new_states.push(pi.into());
                 proofs.push(proof.into());
+
+                let parsed_stake_state = ParsedStakeState {
+                    threshold: field_to_u256(stt.threshold),
+                    bls_key_comm: field_to_u256(stt.stake_table_bls_key_comm),
+                    schnorr_key_comm: field_to_u256(stt.stake_table_schnorr_key_comm),
+                    amount_comm: field_to_u256(stt.stake_table_amount_comm),
+                };
+                stake_stakes.push(parsed_stake_state.into());
```
Comment (philippecamacho) [inline contracts/rust/diff-test/src/main.rs:525]: Looks like `stake_stakes` vector is not returned. Why is that?

File: contracts/src/LightClient.sol (hunk 2 above)
Comment (alysiahuggins) [reply contracts/src/LightClient.sol]: ah yes thank you, done here https://github.com/EspressoSystems/espresso-sequencer/commit/3f4417012c9b75e967b24de3ae35b02b9c7f504b

File: contracts/src/LightClient.sol (hunk 4 above)
Comment (alysiahuggins) [reply contracts/src/LightClient.sol]: ah yes, it was used when we had the mapping, updated here https://github.com/EspressoSystems/espresso-sequencer/commit/56b88cd34eeab65ce00ba3e8a954accfc850937f

File: hotshot-state-prover/src/circuit.rs [hunk 9]
```diff
@@ -68,10 +68,6 @@ impl LightClientStateVar {
                 circuit.create_public_variable(view_number_f)?,
                 circuit.create_public_variable(block_height_f)?,
                 circuit.create_public_variable(state.block_comm_root)?,
-                circuit.create_public_variable(state.fee_ledger_comm)?,
-                circuit.create_public_variable(state.stake_table_comm.0)?,
-                circuit.create_public_variable(state.stake_table_comm.1)?,
-                circuit.create_public_variable(state.stake_table_comm.2)?,
```
Comment (alxiong) [inline hotshot-state-prover/src/circuit.rs:74]: ah this is wrong actually. 🤦 
(i wasted some time assuming our Public input is reduced to 4) 

we should still keep the stake table commitments in the public inputs for security/integrity!!

but we can safely remove `fee_ledger_comm`, so PI.length is 7 now.

cc @mrain @alysiahuggins

File: contracts/rust/diff-test/src/main.rs (hunk 8 above)
Comment (alxiong) [reply contracts/rust/diff-test/src/main.rs]: this will be removed in my future commit of mine

File: hotshot-state-prover/src/circuit.rs (hunk 9 above)
Comment (alxiong) [reply hotshot-state-prover/src/circuit.rs:74]: ah, I need to fix `HotShot` side as well for the struct declaration.

File: contracts/src/LightClient.sol (hunk 1 above)
Comment (alxiong) [reply contracts/src/LightClient.sol:111]: btw, @alysiahuggins i'm addressing my own comments locally. just FYI to avoid redundant work

File: hotshot-state-prover/src/circuit.rs [hunk 10]
```diff
@@ -292,32 +283,32 @@ where
...
         &mut circuit,
         &stake_amount_preimage_vars,
         1,
     )?[0];
-    circuit.enforce_equal(
-        stake_amount_comm,
-        lightclient_state_pub_var
-            .stake_table_comm()
-            .stake_amount_comm,
-    )?;
```
Comment (alxiong) [inline hotshot-state-prover/src/circuit.rs:320]: realized that these are wrong, we need these to ensure security. 

sry that I miss it during my last review.

File: contracts/src/LightClient.sol (hunk 1 above)
Comment (alxiong) [reply contracts/src/LightClient.sol]: done in [f7c5afa](https://github.com/EspressoSystems/espresso-sequencer/pull/1939/commits/f7c5afa0b52200c2e9ac8573ef6ca92c7aed7471)

File: contracts/src/LightClient.sol (hunk 3 above)
Comment (alxiong) [reply contracts/src/LightClient.sol]: addressed in [c63635e](https://github.com/EspressoSystems/espresso-sequencer/pull/1939/commits/c63635e04eb70c651e8bd2d3bf7295e1ee5f9b80)

File: contracts/rust/diff-test/src/main.rs (hunk 8 above)
Comment (alxiong) [reply contracts/rust/diff-test/src/main.rs]: removed in [af51261](https://github.com/EspressoSystems/espresso-sequencer/pull/1986/commits/af5126122179c8c58999c32b3a084b1bd8e4a2c0)

File: hotshot-state-prover/src/circuit.rs (hunk 9 above)
Comment (alxiong) [reply hotshot-state-prover/src/circuit.rs:74]: fixed in [ab81b0b](https://github.com/EspressoSystems/espresso-sequencer/pull/1986/commits/ab81b0bad7497dbcedc4434cd1881cb6891d7c58)

File: hotshot-state-prover/src/circuit.rs (hunk 10 above)
Comment (alxiong) [reply hotshot-state-prover/src/circuit.rs:320]: fixed in [ab81b0b](https://github.com/EspressoSystems/espresso-sequencer/pull/1986/commits/ab81b0bad7497dbcedc4434cd1881cb6891d7c58)

**Review by alxiong — APPROVED** [review]
